### ISSUE TITLE
Selected Node Styling & Node Click Event Handling

### ISSUE THREAD

**Issue by steve-king-ky** [bumbeishvili/org-chart#79]
Firstly, this is a GREAT chart! 

Now, to my questions:

1. Is it possible to subscribe to a click event on the node? 
2. Is it possible to change the styling of a selected node?

**Comment by bumbeishvili** [comment]
Thank you, as for your questions

> 1. Is it possible to subscribe to a click event on the node? 

Yes, there is a special function exposed for that. Unfortunately, it only passes the ID of the clicked node back. I forgot about this method, otherwise, I'd return the whole data object. But now you have to manually loop over your data to find the node you clicked
```
    chart
        .onNodeClick((d, i, arr) => {
          console.log(d, 'Id of clicked node ');
          const obj = <YourData>.filter(obj=>obj.id==d)[0];
        })
```

As for your second question:
> Is it possible to change the styling of a selected node?


Yes, it's possible. Check out how it is done in [this example](https://stackblitz.com/edit/web-platform-o5t1ha).

Basically, after a node is selected, it will have two properties added `_highlighted` and `_upToTheRootHighlighted`. And you can conditionally style the node and links based on those properties inside the following methods
`nodeUpdate` - there is a hidden rectangle (`.node-rect`) which borders every node, you can manipulate this element here

 `nodeContent` - if you want node content itself to be styled conditionally you can manipulate it here

 `linkUpdate` - link manipulation happens here

Check how it is done in linked example

**Comment by steve-king-ky** [comment]
Excellent – thank you!

Steve M. King
Information Systems Manager
Division of Technology Services
Kentucky Personnel Cabinet
***@***.***<mypurpose.ky.gov>
501 High Street | Frankfort, KY 40601
502-564-5529 | ***@***.***
[TEAM_KY_BLUE_sig]
***@***.***<https://personnel.ky.gov/pages/default.aspx>    ***@***.*** <https://www.facebook.com/kypersonnel>  ***@***.*** <https://twitter.com/kypersonnel>  ***@***.*** <https://www.linkedin.com/company/commonwealth-of-kentucky/>  ***@***.*** <https://www.youtube.com/user/PersonnelKY>  ***@***.*** <https://public.govdelivery.com/accounts/KYPERS/subscriber/topics?qsp=CODE_RED>

This communication contains confidential information and is for use by the intended recipient(s) exclusively. If you are not an intended recipient, please note that any form of distribution, copying, forwarding, or use of this communication or the information therein is strictly prohibited and may be unlawful. If you have received this communication in error, please return it to the sender, delete this communication and destroy all copies.

**Comment by steve-king-ky** [comment]
One follow-up question, if I may. How would you recommend lazy-loading? I've got a set of data (org chart) that's got around 7,000 nodes (maybe more in production). I'd rather not load them all at once.

**Comment by bumbeishvili** [comment]
Currently, the org chart is not optimized for lazy loading, I have that feature in mind though for the v3 version (won't be here for months (maybe a year?))

To be honest I am not exactly sure what will be the best way to use it for lazy loading. 

If I had such task probably I would do the following

1. in `onNodeUpdate` - I'd attach new onclick event listener to the node
2. In the event handler, I'd check if we had data for it , if not would make a call and would load its children 
3. I'd update originalData and would redraw the graph
```javascript
chart.data(oldData.concat(newNodes)).render()
```

**Comment by steve-king-ky** [comment]
Awesome idea. Thank you. And could I be so bold as ask a couple of questions? 

That is, is there a way to open the tree at a particular node after it loads and is there a limit on how many nodes that can be loaded or at least a maximum recommended number of nodes?

Thanks again!!

**Comment by bumbeishvili** [comment]
> That is, is there a way to open the tree at a particular node after it loads

Yes, there is a `setCentered` method, which receives nodeId as an argument it centers and expands node. You should rerender the graph to make it work, it just sets the state up. 

https://github.com/bumbeishvili/org-chart/blob/98850af142167ad958b25eb09e78998b1e4b885b/src/d3-org-chart.js#L1282

> Is there a limit on how many nodes can be loaded or at least a maximum recommended number of nodes

There is no limit, browsers can handle about 10 000 nodes in a collapsed state and 2 000 nodes in an expanded state. It very much also depends on the content of nodes

[This example](https://stackblitz.com/edit/web-platform-o5t1ha) has about 1500 nodes and it breaks when all of the nodes are expanded.  But I am sure [this example](https://stackblitz.com/edit/web-platform-sgsxzp) can handle up to 3000 expanded

**Comment by harellevy** [comment]
I guess the best for lazy loading will be to virtual scroll it. maybe save on the node the min,max values for xy of their borders and see if it's in the view + some margins.
you can also make sure not to pass a threshold of default node value (to be exposed in attrs), calculate how many levels can fit, and collapse / aggregate the other nodes to a group with number of direct children and total nodes under. i.e node that contains 5 children and total of 35 including subitems: 5 (35)

**Comment by bumbeishvili** [comment]
> I guess the best for lazy loading will be to virtual scroll it. maybe save on the node the min,max values for xy of their borders and see if it's in the view + some margins.

That's possible, but not sure how well it will align with actual use cases

> you can also expose in attrs nodeThreshold and make sure not to pass that number of nodes , calculate how many levels can fit, and collapse / aggregate the children nodes on lower level to a group with number of direct children and total nodes under. i.e node that contains 5 children and total of 35 including subitems: 5 (35)

I am not sure if I understand this idea well, can you explain it in more detail?
